File: src/tests/csharp/metrics/ErrorMetricsTest.cs
using System;
using NUnit.Framework;
using System.IO;
using Illumina.InterOp.Metrics;
using Illumina.InterOp.Comm;

namespace Illumina.InterOp.Interop.UnitTest
{
	/// <summary>
	/// Confirm that the Error metrics InterOp works properly in C#
	/// </summary>
	[TestFixture]
	public class ErrorMetricsTestV3
	{
		const int Version = 3;
		base_error_metrics expected_metric_set;
		base_error_metrics actual_metric_set = new base_error_metrics();
		vector_error_metrics expected_metrics = new vector_error_metrics();
		byte[] expected_binary_data;

		/// <summary>
		/// Sets up the expected and actual metric sets for the test
		/// </summary>
		[SetUp]
		protected void SetUp()
		{
	        base_cycle_metric_header header = new base_cycle_metric_header();
			expected_metrics.Add(new error_metric(7, 1114, 1, 0.450100899f));
			expected_metrics.Add(new error_metric(7, 1114, 2, 0.900201797f));
			expected_metrics.Add(new error_metric(7, 1114, 3, 0.465621591f));

			int[] tmp = new int[]{3,30,7,0,90,4,1,0,-96,115,-26,62,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
				,7,0,90,4,2,0,-96,115,102,63,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
				,7,0,90,4,3,0,-12,101,-18,62,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
			};
			expected_binary_data = new byte[tmp.Length];
			for(int i=0;i<expected_binary_data.Length;i++)
			{
			    expected_binary_data[i] = (byte)tmp[i];
			}
			expected_metric_set = new base_error_metrics(expected_metrics, Version, header);
			c_csharp_comm.read_interop_from_buffer(expected_binary_data, (uint)expected_binary_data.Length, actual_metric_set);
		}

		/// <summary>
		/// Confirms that the data was properly parsed and matches the expected model.
		/// This test also confirms that the binding gives the expected results.
		/// </summary>
		[Test]
		public void TestHardCodedBinaryData()
		{
			Assert.AreEqual(expected_metric_set.version(),  actual_metric_set.version());
			Assert.AreEqual(expected_metric_set.size(),  actual_metric_set.size());

			for(uint i=0;i<Math.Min(expected_metric_set.size(), actual_metric_set.size());i++)
			{
				Assert.AreEqual(expected_metric_set.at(i).lane(), actual_metric_set.at(i).lane());
				Assert.AreEqual(expected_metric_set.at(i).tile(), actual_metric_set.at(i).tile());
				Assert.AreEqual(expected_metric_set.at(i).cycle(), actual_metric_set.at(i).cycle());
				Assert.AreEqual(expected_metric_set.at(i).errorRate(), actual_metric_set.at(i).errorRate(), 1e-7);
				Assert.AreEqual(expected_metric_set.at(i).mismatch_count(), actual_metric_set.at(i).mismatch_count());
				for(uint j=0;j<expected_metric_set.at(i).mismatch_count();j++)
					Assert.AreEqual(expected_metric_set.at(i).mismatch_cluster_count(j), actual_metric_set.at(i).mismatch_cluster_count(j));
			}
            byte[] newBuffer = new byte[c_csharp_comm.compute_buffer_size(expected_metric_set)];
			c_csharp_comm.write_interop_to_buffer(expected_metric_set, newBuffer, (uint)newBuffer.Length);
			Assert.AreEqual(newBuffer, expected_binary_data);
		}
	}
}

File: src/tests/interop/metrics/error_metrics_test.cpp
/** Unit tests for the error metrics
 *
 *
 *  @file
 *  @date 8/23/2015
 *  @version 1.0
 *  @copyright GNU Public License.
 */

#ifdef _MSC_VER
#pragma warning(disable:4127) // MSVC warns about using constants in conditional statements, for template constants
#endif

#include <gtest/gtest.h>
#include "interop/model/run_metrics.h"
#include "src/tests/interop/metrics/inc/error_metrics_test.h"
#include "src/tests/interop/inc/generic_fixture.h"
#include "src/tests/interop/inc/proxy_parameter_generator.h"
#include "src/tests/interop/metrics/inc/metric_generator.h"
using namespace illumina::interop::model::metrics;
using namespace illumina::interop::model::metric_base;
using namespace illumina::interop::io;
using namespace illumina::interop;
using namespace illumina::interop::unittest;

typedef metric_set< error_metric > error_metric_set;
/** Setup for tests that compare two error metric sets */
struct error_metrics_tests : public generic_test_fixture< error_metric_set > {};

error_metrics_tests::generator_type error_unit_test_generators[] = {
        wrap(new hardcoded_metric_generator< error_metric_v3 >) ,
        wrap(new hardcoded_metric_generator< error_metric_v4 >),
        wrap(new hardcoded_metric_generator< error_metric_v5 >),
        wrap(new hardcoded_metric_generator< error_metric_v6 >),
        wrap(new write_read_metric_generator< error_metric_v3 >),
        wrap(new write_read_metric_generator< error_metric_v4 >),
        wrap(new write_read_metric_generator< error_metric_v5 >),
        wrap(new write_read_metric_generator< error_metric_v6 >),
        wrap(new by_cycle_metric_generator< error_metric_v3 >),
        wrap(new by_cycle_metric_generator< error_metric_v4 >),
        wrap(new by_cycle_metric_generator< error_metric_v5 >),
        wrap(new by_cycle_metric_generator< error_metric_v6 >),
        wrap(new clear_metric_generator< error_metric_v3 >),
        wrap(new clear_metric_generator< error_metric_v4 >),
        wrap(new clear_metric_generator< error_metric_v5 >),
        wrap(new clear_metric_generator< error_metric_v6 >)
};

// Setup unit tests for error_metrics_tests
INSTANTIATE_TEST_CASE_P(error_metric_unit_test,
                        error_metrics_tests,
                        ::testing::ValuesIn(error_unit_test_generators));

/**
 * @class illumina::interop::model::metrics::error_metric
 * @test Confirm version 3 of the metric can be written to and read from a stream
 * @test Confirm version 3 of the metric matches known binary file
 */
TEST_P(error_metrics_tests, compare_expected_actual)
{
    if(skip_test) return;
    ASSERT_TRUE(fixture_test_result);
    ASSERT_EQ(actual.version(), expected.version());
    ASSERT_EQ(actual.size(), expected.size());
    EXPECT_EQ(actual.max_cycle(), expected.max_cycle()) << actual.offset_map().size()
                                                        << " == "
                                                        << expected.offset_map().size();

    for(value_type::const_iterator it_expected=expected.begin(), it_actual = actual.begin();
        it_expected != expected.end();
        it_expected++,it_actual++)
    {
        EXPECT_EQ(it_expected->lane(), it_actual->lane());
        EXPECT_EQ(it_expected->tile(), it_actual->tile());
        EXPECT_EQ(it_expected->cycle(), it_actual->cycle());
        EXPECT_EQ(it_expected->mismatch_count(), it_actual->mismatch_count());
        EXPECT_NEAR(it_expected->error_rate(), it_actual->error_rate(), 1e-5f);
        INTEROP_EXPECT_NEAR(it_expected->phix_adapter_rate(), it_actual->phix_adapter_rate(), 1e-5f);
        ASSERT_EQ(it_expected->phix_adapter_count(), it_actual->phix_adapter_count());
        for(size_t i=0;i<it_expected->phix_adapter_count();++i)
            INTEROP_EXPECT_NEAR(it_expected->phix_adapter_rate_at(i), it_actual->phix_adapter_rate_at(i), 1e-5f);

        for(ptrdiff_t i=0;i<static_cast<ptrdiff_t>(it_expected->mismatch_count());i++)
            EXPECT_EQ(it_expected->mismatch_cluster_count(i), it_actual->mismatch_cluster_count(i));
    }
}

/**
 * @test Ensure tile_numbers_for_lane returns the proper number
 */
TEST(error_metrics_single_test, test_tile_metric_count_for_lane)
{
    error_metric_set metrics;
    error_metric_v3::create_expected(metrics);
    EXPECT_EQ(metrics.tile_numbers_for_lane(7).size(), 1u);
}


/**
 * @test Ensure the adapter sequences are correct in header
 */
TEST(error_metrics_single_test, test_error_header_v6)
{
    error_metric_set metrics;
    error_metric_v6::create_expected(metrics);
    EXPECT_EQ(metrics.number_adapters(), 2u);
    EXPECT_EQ(metrics.adapter_length(), 8u);
    EXPECT_EQ(metrics.adapter_sequences().size(), 2u);
    EXPECT_EQ(metrics.adapter_sequences()[0], "CCCCCCCC");
    EXPECT_EQ(metrics.adapter_sequences()[1], "AAAAAAAA");
}

/**
 * @test Ensure the rates per sequences are correct
 */
TEST(error_metrics_single_test, test_error_rates_v6)
{
    error_metric_set metrics;
    error_metric_v6::create_expected(metrics);

    float rates[3][2] = {{0.298748f, 0.1234f},
                         {0.287257f, 0.3934f},
                         {0.608985f, 0.5034f}};

    EXPECT_EQ(metrics.size(), 3u);
    size_t i = 0;
    for(error_metric_v6::const_iterator it=metrics.begin(); it != metrics.end();it++, i++) {
        EXPECT_EQ(it->phix_adapter_rate(), (rates[i][0]+rates[i][1])/2);
    }
}


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Setup regression test
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
regression_test_metric_generator<error_metric_set> error_regression_gen;
INSTANTIATE_TEST_CASE_P(error_metric_regression_test,
                        error_metrics_tests,
                        ProxyValuesIn(error_regression_gen, regression_test_data::instance().files()));




